### ISSUE TITLE
Failing to build a jitclass with a deferred_type typed container as class member

### ISSUE THREAD

**Issue by remydubois** [numba/numba#6916]
Hello,

thanks for maintaining and developing Numba.

I encountered difficulties when trying to use typed containers (list) as attribute of a jit class, when the container is typed to a deferred type, see example below:

```python
# Assigning defer-type typed container as class member fails
from collections import OrderedDict
from numba import types, typed, deferred_type
from numba.experimental import jitclass

specs = OrderedDict()
node_type = deferred_type()
specs['children'] = types.ListType(node_type)
@jitclass(specs)
class Node:
    def __init__(self):
        self.children = typed.List.empty_list(node_type)

node_type.define(Node.class_type.instance_type)
# this will fail:
Node()
```

The exact error is:

> Assertion failed: (Ty->isSized() && "Cannot getTypeInfo() on a type that is unsized!"), function getTypeSizeInBits, file /Users/ci/miniconda3/conda-bld/llvmdev_1601977477900/work/include/llvm/IR/DataLayout.h, line 625.
Fatal Python error: Aborted

Note:
I successfully compiled a jitclass with a typed container, as the documentation explains:
```python
# Assigning builtin type typed container as class member fails
specs = OrderedDict()
node_type = deferred_type()
specs['children'] = types.ListType(int32)
@jitclass(specs)
class Node:
    def __init__(self):
        self.children = typed.List.empty_list(int32)

node_type.define(Node.class_type.instance_type)
# this will work:
Node()
```
Equally, I managed to compile jitclass with defer type typed class member (not containerized)
```python
specs = OrderedDict()
node_type = deferred_type()
specs['children'] = optional(node_type)
@jitclass(specs)
class Node:
    def __init__(self):
        self.children = None

node_type.define(Node.class_type.instance_type)
# Create node and assign children works:
Node().children = Node()
```
All I wish is a combination of both.

Context: my goal is to build a tree where the amount of children per node depends on the dimensionality of the data. It is thus difficult to "hard code" the amount of children to declare as attribute.

I am using numba 0.53.1 on a macos platform.

Thank you for your help,

**Comment by stuartarchibald** [comment]
Thanks for the report. I think the root cause of this error is likely the same as/is a duplicate of https://github.com/numba/numba/issues/6077.

**Comment by remydubois** [comment]
Thanks for looking into this,
Do you know if there could be ways to support this feature in the future ?

**Comment by tonyabracadabra** [comment]
I am recently running into the same issue, is there any feature update/proposals on this thread? this can be really useful for implementing complex data structures with numba

**Comment by esc** [comment]
> I am recently running into the same issue, is there any feature update/proposals on this thread? this can be really useful for implementing complex data structures with numba

No, any updates to this thread would be in this thread. Unfortunately there are no updates at this time!
